### PR TITLE
Ensure the first demo is fully successful

### PR DESCRIPTION
**Pull request by russellb**
commit ca23a49bbd5e0ccc010ab2edca381cb396df1309 (main)
Author: Russell Bryant <russell.bryant@gmail.com>
Date:   Wed Feb 28 10:49:59 2024 -0500

    Ensure the first demo is fully successful
    
    `intro-demo` from chapter1 regularly does not reach 100% accuracy in
    my testing. The text in the notebook seems to imply this is because of
    the size of the input. However, given that the input contains the
    entirety of the problem space, I believed we should be able to
    reliably provide a demo that produces 100% accuracy.
    
    I tried a number of changes, but settled on increasing the size of the
    hidden layer as having the most significant impact on the result. By
    going from a size of 4 to 32, I now get 100% accuracy reliably with
    less than 200 epochs, and usually far fewer than that. The code now
    does it 10 epochs at a time and stops when it's at 100%.
    
    I've also dropped the text that tried to explain why the accuracy may
    be low, as we should now always have an accurate result.
    
    This was a helpful post where someone else explained the steps they
    went through for the same example and which variables produced the
    most impact for them.
    
        https://blog.thoughtram.io/machine-learning/2016/11/02/understanding-XOR-with-keras-and-tensorlow.html
    
    Many thanks to Chris Sibbit who helped me understand many details
    about neutral network structure, loss functions, optimization
    algorithms, and debugging techniques based on his experience.
    
    Co-authored-by: Chris Sibbitt <csibbitt@redhat.com>
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 68fedec54eb4dce10e0d01ffcc7cf3aed602d92d (HEAD -> chapter1-intro-demo, russellb/chapter1-intro-demo)
Author: Russell Bryant <russell.bryant@gmail.com>
Date:   Wed Feb 28 14:21:53 2024 -0500

    chapter1: Adjust loss function and optimization algorithm
    
    This change adjusts the first demo notebook to use a different loss
    function and optimization algorithm.
    
    The optimization algorithm is changed from `adam` to `rmsprop`. Its
    simpler and easier to understand, plus it seems to converge at 100%
    accuracy in 20% to 30% fewer epochs than `adam`. It also appears to
    run faster, though I didn't measure that directly.
    
    The loss function is changed from `binary_crossentropy` to
    `mean_squared_error`. Similar to the optimization algorithm, we
    benefit from choosing a simpler and easier to use function for our
    first example. `binary_crossentropy` is also discussed as being best
    for binary classification problems. We aren't trying to determine the
    probabiliy among two classifications here, so the simpler
    `mean_squared_error` loss function seems more appropriate. Over 100
    runs, they seem to converge on 100% accuracy in a similar number of
    epochs.
    
    Co-authored-by: Chris Sibbitt <csibbitt@redhat.com>
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

### CODE REVIEW

**Comment by russellb** [comment]
cc @csibbitt

File: 1.intro/chapter1/intro-demo/notebook.ipynb [hunk 1]
```diff
@@ -164,23 +176,28 @@
...
     "\n",
     "# Compile the model\n",
     "model.compile(loss=\"binary_crossentropy\", optimizer='adam', metrics=['accuracy'])\n",
     "\n",
-    "# Train the model\n",
-    "model.fit(inputs, output, epochs=1000)"
+    "# Train the model until it reaches 100% accuracy. Given the simple XOR use case and that our training\n",
+    "# data encompasses the entire problem space of possible inputs and outputs, we know we can get it to 100%.\n",
+    "epochs = 0\n",
+    "while model.fit(inputs, output, epochs=epochs+10, initial_epoch=epochs).history['accuracy'][-1] != 1.0:\n",
```
Comment (csibbitt) [inline 1.intro/chapter1/intro-demo/notebook.ipynb:199]: Interesting approach!

File: 1.intro/chapter1/intro-demo/notebook.ipynb [hunk 2]
```diff
@@ -164,23 +176,28 @@
...
+   },
    "outputs": [],
    "source": [
     "from keras.layers import Dense\n",
     "from keras.models import Sequential\n",
     "\n",
     "# Define a sequential neural network\n",
     "model = Sequential([\n",
-    "    Dense(4, input_dim=2, activation=\"relu\"),\n",
+    "    Dense(32, input_dim=2, activation=\"relu\"),\n",
```
Comment (csibbitt) [inline 1.intro/chapter1/intro-demo/notebook.ipynb:189]: :+1:

File: 1.intro/chapter1/intro-demo/notebook.ipynb [hunk 3]
```diff
@@ -164,23 +176,28 @@
...
     "    Dense(1, activation=\"sigmoid\")\n",
     "])\n",
     "\n",
     "# Compile the model\n",
     "model.compile(loss=\"binary_crossentropy\", optimizer='adam', metrics=['accuracy'])\n",
```
Comment (csibbitt) [inline 1.intro/chapter1/intro-demo/notebook.ipynb]: The use of binary_crossentropy is still suspicious to me because the output layer does not represent classification probabilities.  A simpler regression loss like mean_squared_error seems more appropriate. (Whether that actually brings any performance benefit is a different question)

**Review by csibbitt — APPROVED** [review]

File: 1.intro/chapter1/intro-demo/notebook.ipynb (hunk 3 above)
Comment (russellb) [reply 1.intro/chapter1/intro-demo/notebook.ipynb]: It seems to perform just as well. Given that this is demo 1, I think using the simplest algorithms makes sense if they work fine. Using any of the more advanced things complicates the example more than necessary. That plus your point that it may not be ideal anyway makes a strong case to change it! I can update the PR with an adjustment here, as well. I will also try to update the commit message and PR description to explain the change.

Comment (russellb) [reply 1.intro/chapter1/intro-demo/notebook.ipynb]: @csibbitt I've added a second commit to the PR, adjusting the loss function and optimization algorithm. The commit message (and PR description) include the justification based on our discussion.

**Review by jramcast — APPROVED** [review]
Thank you very much for this contribution @russellb @csibbitt . Approved and merged :+1:
